packages/plugin-npm-cli/sources/commands/npm/audit.ts

```diff
@@ -0,0 +1,217 @@
+    `,
+    examples: [[
+      `Checks for known security issues with the installed packages. The output is a list of known issues.`,
+      `yarn npm audit`,
+    ], [
+      `Output moderate (or more severe) vulnerabilities`,
+      `yarn npm audit --severity moderate`,
+    ], [
+      `Show audit report as valid JSON`,
+      `yarn npm audit --json`,
+    ]],
+  });
+
+  @Command.Path(`npm`, `audit`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+    const {project, workspace} = await Project.find(configuration, this.context.cwd);
+
+    if (!workspace)
+      throw new WorkspaceRequiredError(project.cwd, this.context.cwd);
+
+    const report = await StreamReport.start({
+      configuration,
+      includeFooter: false,
+      json: this.json,
+      stdout: this.context.stdout,
+    }, async report => {
+      const requiredDependencies = project.workspaces.reduce((acc, workspace) =>
+        [...acc, ...workspace.manifest.dependencies.values()], new Array());
+      const requiredDevDependencies = project.workspaces.reduce((acc, workspace) =>
+        [...acc, ...workspace.manifest.devDependencies.values()], new Array());
+      const requiredAllDependencies = [...requiredDependencies, ...requiredDevDependencies];
+      const requires = transformDescriptorIterableToRequiresObject(requiredAllDependencies);
+
+      const transitiveDevDependencies = getTransitiveDevDependencies(project, workspace);
+      const dependencies = Array.from(project.originalPackages.values()).reduce((acc, cur) => ({
+        ...acc, [cur.scope ? `@${cur.scope}/${cur.name}` : cur.name]: {
+          version: cur.version,
+          integrity: cur.identHash,
+          requires: transformDescriptorIterableToRequiresObject(cur.dependencies.values()),
+          dev: transitiveDevDependencies.has(cur.identHash),
+        },
+      }), {});
+
+      const body = {
+        requires,
+        dependencies,
+      };
+      const registry = npmConfigUtils.getPublishRegistry(workspace.manifest, {configuration});
+
+      let result: AuditResponse;
+      try {
+        const url = `/-/npm/v1/security/audits`;
+        result = (await npmHttpUtils.post(url, body, {
+          authType: npmHttpUtils.AuthType.NO_AUTH,
+          configuration,
+          jsonResponse: true,
+          registry,
+        }) as unknown) as AuditResponse;
+      } catch (err) {
+        if (err.name !== `HTTPError`) {
+          throw err;
+        } else {
+          throw new ReportError(MessageName.EXCEPTION, err.toString());
+        }
+      }
+
+      report.reportJson(result);
+
+      if (isError(result.metadata.vulnerabilities, this.severity)) {
+        report.reportError(MessageName.EXCEPTION, JSON.stringify(result, undefined, 2));
```

Comment:
I haven't added any pretty outputs yet. Right now if there is an error it returns the whole JSON report, and if it passes it exits without printing.
Should I simply add output here to match existing behaviour in yarn classic?

Also wasn't sure if I needed to create a new exception code, or if I can reuse the generic `EXCEPTION`.

If there are multiple vulnerabilities, should they each be their own individual error in the report, or a single error?

Reply:
We'll need the default display to be pretty, and the JSON stream to be returned with the `--json` flag (to be consistent with the other commands). We don't need to (and we probably shouldn't) follow the output from Yarn Classic - I prefer to careful craft an appealing output without being burdened by the past decisions.

In this particular case, I think the best would be to use the new `emitTree` function from `treeUtils` in order to print the tree of 'vulnerable packages' (with only the associated CVEs), and perhaps to provide more details when the `-v,--verbose` flag is set.

Reply:
Cheers @arcanis. I did implement the `--json` flag and used the `reportJson` function just above this, so that should all be working. Just didn't want to touch the human readable output until I had an idea what you guys were after.

Will checkout the `emitTree` function. Thanks for the pointer 👍